versioned_hdf5/staged_changes.py

```diff
@@ -394,12 +427,42 @@ def _load_plan(self, copy: bool = True) -> LoadPlan:
         )
 
     def _get_slab(
-        self, idx: int | None, default: NDArray[T] | None = None
+        self,
+        idx: int | None,
+        default: NDArray[T] | None = None,
+        writeable: bool = False,
     ) -> NDArray[T]:
+        """Fetch a slab by index, or the __getitem__/__setitem__ array if idx is None.
+
+        If there was a previous call to copy() and the slab is going to be mutated,
+        eagerly perform a deep-copy of the slab before returning it.
+        If there was a previous call to astype(), eagerly convert the slab to the new
+        dtype. In both cases, the returned slab replaces the previous one in self.slabs.
+        """
         slab = self.slabs[idx] if idx is not None else default
         assert slab is not None
-        assert slab.dtype == self.dtype, (slab.dtype, self.dtype)
         assert slab.ndim == self.ndim
+
+        if slab.dtype != self.dtype:
+            # There was a previous call to astype().
+            # `slab` Must be a staged slab.
+            # The full slab and __setitem__ value are converted eagerly.
+            # Base slabs are eagerly loaded or swapped out.
+            assert idx is not None
+            assert idx > self.n_base_slabs  # only staged slabs are converted lazily
+            slab = asarray(slab, dtype=self.dtype)
+            self.slabs[idx] = slab
+
+        if writeable:
+            # dst_slab is either a staged slab or the return value of __getitem__
+            assert isinstance(slab, np.ndarray)
+            if not slab.flags.writeable:
+                assert idx is not None  # __getitem__ return value is always writeable
+                assert idx > self.n_base_slabs  # Only staged slabs are writeable
+                # There was a previous call to copy()
+                slab = slab.copy()
+                self.slabs[idx] = slab
```

Comment:
Are there instances in which both of these would be hit, resulting in an extra (needless) copy of the slab? For example if both `astype()` and `copy()` were called previously?

Reply:
When `asarray` performs a deep copy in the 451:454 branch, the returned array is always writeable, so the second branch 460:464 is not executed.

There is one special case in `asarray` for uint<->int conversion, which is performed as a view. In that case, you may hit both branches: first `asarray` will return a read-only uint view of a read-only int buffer (or vice versa), which is cheap, then `.copy()` will deep-copy the buffer, which is the only expensive bit.